File: smallest_divisible_product1.c
// 3345. Smallest Divisible Digit Product I
#include "leetcode.h"

/*
 * you are given two integers 'n' and 't'. return the smallest nmber greater
 * than or equal to 'n' such that the product of its digits is divisible by 't'.
 */

int digit_product(int n) {
  int res = 1;
  while (n > 0) {
    int rem = n % 10;
    res *= rem;
    n /= 10;
  }
  return res;
}

int smallestNumber(int n, int t) {
  while (true) {
    int val = digit_product(n);
    if (!(val % t))
      return n;
    n++;
  }
}

int main() {
  int r1 = smallestNumber(10, 2);
  int r2 = smallestNumber(15, 3);
  printf("%d\n", r1);
  assert(r1 == 10);
  printf("%d\n", r2);
  assert(r2 == 16);
}

File: smallest_divisible_product1.py
# 3345. Smallest Divisible Digit Product I
from math import gcd

"""
you are given two integers 'n' and 't'. return the smallest nmber greater
than or equal to 'n' such that the product of its digits is divisible by 't'.
"""


class Solution:
    def smallestNumber(self, n: int, t: int) -> int:
        q, r = divmod(n, 10)
        req = t // gcd(max(q, 1), t)
        nxt = ((r + req - 1) // req) * req
        x = nxt - (nxt - 10) * (nxt // 10)
        return q * 10 + x


if __name__ == "__main__":
    obj = Solution()
    print(obj.smallestNumber(n=10, t=2))
    print(obj.smallestNumber(n=15, t=3))
